feat: Add Enterprise Managed Authorization (SEP-990) support (#1305)
modelcontextprotocol/csharp-sdk: README.md

- [Protocol Specification](https://modelcontextprotocol.io/specification/)
- [GitHub Organization](https://github.com/modelcontextprotocol)

> ## Enterprise Auth / Enterprise Managed Authorization

Reviewer: ```suggestion
## Cross-Application Access (Identity Assertion Authorization Grant flow)
```

I think we should de-emphasize "Enterprise" now that it's no longer part of the type names. This appears to align better with the TS SDK's conceptual docs.

https://github.com/modelcontextprotocol/typescript-sdk/blob/48251fe2cb24fb9b2a12ef3cffb5f5bd196a7ea4/docs/client.md#L169

The SDK provides Cross-Application Access support for the [Identity Assertion Authorization Grant flow](https://github.com/modelcontextprotocol/ext-auth/blob/main/specification/draft/enterprise-managed-authorization.mdx),
enabling enterprise SSO scenarios where users authenticate once via their enterprise Identity Provider and
access MCP servers without per-server authorization prompts.

The flow consists of two token operations:
1. **RFC 8693 Token Exchange** at the IdP: ID Token → JWT Authorization Grant (JAG)
2. **RFC 7523 JWT Bearer Grant** at the MCP Server: JAG → Access Token

### Using CrossApplicationAccessProvider

```csharp
using ModelContextProtocol.Authentication;

var provider = new CrossApplicationAccessProvider(new CrossApplicationAccessProviderOptions
{
    ClientId = "mcp-client-id",
    IdpTokenEndpoint = "https://company.okta.com/oauth2/token",
    IdpClientId = "idp-client-id",
    IdTokenCallback = async (context, ct) =>
    {
        // Return the OIDC ID token from your SSO session

>         return myIdToken;

Reviewer: This callback is `async` but has no `await`, which will compile-warn (CS1998). Either drop `async` and return `Task.FromResult(myIdToken)`, or make the example realistically asynchronous by showing the ID-token fetch:

```suggestion
        // Fetch a fresh ID token from your SSO session.
        return await mySsoClient.GetIdTokenAsync(cancellationToken);
```

    }
});

var tokens = await provider.GetAccessTokenAsync(
    resourceUrl: new Uri("https://mcp-server.example.com"),
    authorizationServerUrl: new Uri("https://auth.mcp-server.example.com"));
```

> 

Reviewer: I thought this was in a conceptual doc. This doubles the size of our README. We try to keep this pretty slim. I would move this to somewhere under the `docs/` directory. We don't have client OAuth documented yet, and I'm not going to ask you to do that. Feel free to add it as a section to [transports.md](https://github.com/modelcontextprotocol/csharp-sdk/blob/712a06b83a656fedf854b8a985f4dcd53d334ea4/docs/concepts/transports/transports.md#streamable-http-transport), and we'll move it to a dedicated client OAuth doc once we create that.

If you want to go the extra mile, you could add this flow to [ProtectedMcpServer](https://github.com/modelcontextprotocol/csharp-sdk/tree/712a06b83a656fedf854b8a985f4dcd53d334ea4/samples/ProtectedMcpServer) and [ProtectedMcpClient](https://github.com/modelcontextprotocol/csharp-sdk/tree/712a06b83a656fedf854b8a985f4dcd53d334ea4/samples/ProtectedMcpClient). These both run against the TestOAuthServer that you've already updated, so I think you're already most of the way there. It could be helpful for referencing from the `/docs` too. We already do that in `identity.md` for the server side at least.

- [Protocol Specification](https://modelcontextprotocol.io/specification/)
- [GitHub Organization](https://github.com/modelcontextprotocol)

> ## Enterprise Auth / Enterprise Managed Authorization (SEP-990)

Reviewer: ```suggestion
## Enterprise Auth / Enterprise Managed Authorization
```

> The SDK provides Enterprise Auth utilities for the Identity Assertion Authorization Grant flow (SEP-990),

Reviewer: ```suggestion
The SDK provides Enterprise Auth utilities for the [Identity Assertion Authorization Grant flow](https://github.com/modelcontextprotocol/ext-auth/blob/main/specification/draft/enterprise-managed-authorization.mdx),
```

        ClientId = "idp-client-id",
    });

// Step 2: Exchange JAG for an access token at the MCP authorization server
var tokens = await EnterpriseAuth.ExchangeJwtBearerGrantAsync(
    new ExchangeJwtBearerGrantOptions
    {
        TokenEndpoint = "https://auth.mcp-server.example.com/token",
        Assertion = jag,
        ClientId = "mcp-client-id",
    });
```

### Using the EnterpriseAuthProvider (Layer 3)

```csharp
var provider = new EnterpriseAuthProvider(new EnterpriseAuthProviderOptions
{
    ClientId = "mcp-client-id",
    AssertionCallback = async (context, ct) =>
    {
        return await EnterpriseAuth.DiscoverAndRequestJwtAuthorizationGrantAsync(
            new DiscoverAndRequestJwtAuthGrantOptions
            {
                IdpUrl = "https://company.okta.com",
                Audience = context.AuthorizationServerUrl.ToString(),
                Resource = context.ResourceUrl.ToString(),
                IdToken = myIdToken,
                ClientId = "idp-client-id",
            }, ct);
    }
});

var tokens = await provider.GetAccessTokenAsync(
    resourceUrl: new Uri("https://mcp-server.example.com"),

>     authorizationServerUrl: new Uri("https://auth.mcp-server.example.com"));

Reviewer: Makes sense. Are APIs that just provide the access token that we can then use to create a new client sufficient? I know for the normal OAuth flows we need a bit deeper integration to support scope step up and token refresh. I suppose scope step up would be hard without an interactive user and a browser, but refresh seems like it could still be useful. I suppose you could always refresh manually and call [ResumeSessionAsync](https://modelcontextprotocol.github.io/csharp-sdk/api/ModelContextProtocol.Client.McpClient.html#ModelContextProtocol_Client_McpClient_ResumeSessionAsync_ModelContextProtocol_Client_IClientTransport_ModelContextProtocol_Client_ResumeClientSessionOptions_ModelContextProtocol_Client_McpClientOptions_Microsoft_Extensions_Logging_ILoggerFactory_System_Threading_CancellationToken_) with the new access token. What do the other SDKs do for token refresh in this case?
